libsolidity/codegen/ir/IRGeneratorForStatements.cpp

```diff
@@ -776,10 +805,42 @@ bool IRGeneratorForStatements::visit(BinaryOperation const& _binOp)
 {
 	setLocation(_binOp);
 
-	solAssert(!!_binOp.annotation().commonType);
+	if (FunctionDefinition const* function = _binOp.annotation().userDefinedFunction)
+	{
+		_binOp.leftExpression().accept(*this);
+		_binOp.rightExpression().accept(*this);
+		setLocation(_binOp);
+
+		solAssert(
+			dynamic_cast<SourceUnit const*>(function->scope()),
+			"Only file-level functions and library functions can be bound to a user type operator."
+		);
+
+		FunctionType const* functionType = dynamic_cast<FunctionType const*>(
+			function->libraryFunction() ? function->typeViaContractName() : function->type()
+		);
+		solAssert(functionType);
+		functionType = dynamic_cast<FunctionType const&>(*functionType).asBoundFunction();
+		solAssert(functionType);
+
+		string left = expressionAsType(_binOp.leftExpression(), *functionType->selfType());
+		string right = expressionAsType(_binOp.rightExpression(), *functionType->parameterTypes().at(0));
+		solAssert(!left.empty() && !right.empty());
+
+		solAssert(function->isImplemented(), "");
+
+		define(_binOp) <<
+			m_context.enqueueFunctionForCodeGeneration(*function) <<
+			("(" + left + ", " + right + ")\n");
+
```

Comment:
Do we need to perform any type conversion of the return value?
We can assume that the value has been converted to the return type of the function, but are we sure it is the same as the type of `_binOp`? Maybe assert that.